components/script/dom/xmlhttprequest.rs

```diff
@@ -796,22 +796,26 @@ impl XMLHttpRequest {
     }
 
     fn process_headers_available(&self, cors_request: Option<CORSRequest>,
-                                 gen_id: GenerationId, metadata: Metadata) -> Result<(), Error> {
-
+                                 gen_id: GenerationId, metadata: Result<Metadata, NetworkError>)
+                                 -> Result<(), Error> {
         if let Some(ref req) = cors_request {
-            match metadata.headers {
-                Some(ref h) if allow_cross_origin_request(req, h) => {},
-                _ => {
-                    self.process_partial_response(XHRProgress::Errored(gen_id, Error::Network));
-                    return Err(Error::Network);
-                }
+            match metadata {
+                Ok(meta) => {
+                    match meta.headers {
+                        Some(ref h) if allow_cross_origin_request(req, h) => (),
+                        _ => {
+                            self.process_partial_response(XHRProgress::Errored(gen_id, Error::Network));
+                            return Err(Error::Network);
+                        }
+                    }
+
+                    // XXXManishearth Clear cache entries in case of a network error
+                    self.process_partial_response(XHRProgress::HeadersReceived(gen_id, meta.headers, meta.status));
+                },
+                Err(_) => return Err(Error::Network),
             }
         }
 
-        // XXXManishearth Clear cache entries in case of a network error
```

Comment:
This change looks incorrect to me, since it no longer is executed in the !cors_request case.